File: coinbase_exchange.py
"""Safe Coinbase exchange-class selection."""

import ccxt


def resolve_coinbase_exchange_class(use_sandbox: bool, ccxt_module=ccxt):
    """Return a Coinbase ccxt class without crossing environment boundaries."""
    if use_sandbox:
        exchange_class = getattr(ccxt_module, "coinbaseexchange", None)
        if exchange_class is None:
            raise RuntimeError(
                "Coinbase sandbox requires ccxt.coinbaseexchange; refusing to "
                "fall back to the production Coinbase API. Update ccxt."
            )
        return exchange_class, "coinbaseexchange"

    for exchange_id in ("coinbase", "coinbaseadvanced", "coinbaseexchange"):
        exchange_class = getattr(ccxt_module, exchange_id, None)
        if exchange_class is not None:
            return exchange_class, exchange_id

    raise RuntimeError("No Coinbase exchange class found in ccxt. Update ccxt.")

File: coinbase_exchange.js
/**
 * Resolve the ccxt Coinbase implementation for the requested environment.
 *
 * Current ccxt versions expose Advanced Trade as `coinbase`. Older releases
 * used `coinbaseadvanced`. Sandbox must remain on `coinbaseexchange`, the only
 * Coinbase implementation with Exchange sandbox endpoints.
 */
export function resolveCoinbaseExchange(ccxtNamespace, useSandbox) {
    const candidateIds = useSandbox
        ? ['coinbaseexchange']
        : ['coinbase', 'coinbaseadvanced', 'coinbaseexchange'];

    for (const exchangeId of candidateIds) {
        const ExchangeClass = ccxtNamespace[exchangeId];
        if (typeof ExchangeClass === 'function') {
            return { ExchangeClass, exchangeId };
        }
    }

    throw new Error(
        `No Coinbase ${useSandbox ? 'sandbox' : 'production'} exchange class found in ccxt. ` +
        'Update ccxt with: npm install ccxt@latest'
    );
}
